doc/clangsay.6
.TH CLANGSAY "6" "2015年11月" "ユーザコマンド"
.SH 名前
clangsay \- マルチバイト文字対応のC言語実装版cowsayクローン
.SH 書式
.B clangsay
[\fIオプション\fR]...
.br
.B clangsay
[\fIオプション\fR]... [\fI文字列\fR]...
.br

.B clangsay
\fI--file=COWFILE\fR [\fIオプション\fR]...
.br
.B clangsay
\fI--file=COWFILE\fR [\fIオプション\fR]... [\fI文字列\fR]...

.SH 説明
.PP
マルチバイト文字に対応したcowsay互換のC言語実装プログラム。
.br
標準入力または、コマンドライン引数として与えられた文字列を、
.br
様々なアスキーアートのテーマファイル(以下COWFILEと表記)と共に標準出力へ出力します。

.br
長いオプションに必須の引数は短いオプションにも必須です。
.TP
\fB\-b\fR, \fB\-\-borg\fR
\&スタートレックのボーグモード。牛の目がooから==になる
.TP
\fB\-d\fR, \fB\-\-dead\fR
\&死体モード。牛の目がXXになり、下を突き出したようにUが口元から伸びる
.TP
\fB\-g\fR, \fB-\-greedy\fR
\&貪欲モード。牛の目が$$になる
.TP
\fB\-p\fR, \fB-\-paranoid\fR
\&偏執狂モード。目が@@になる
.TP
\fB\-s\fR, \fB-\-stoned\fR
\&酔っぱらいモード。牛の目が**になり、下を突き出したようにUが口元から伸びる
.TP
\fB\-t\fR, \fB-\-tired\fR
\&お疲れモード。牛の目が--になる
.TP
\fB\-w\fR, \fB-\-wired\fR
\&興奮モード。牛の目がOOとなり大きく見開く
.TP
\fB\-y\fR, \fB-\-youthful\fR
\&若返りモード。牛の目が..になる
.TP
\fB\-e\fR, \fB-\-eye\fR\fI=EYE\fR
\&牛の目に使う文字列(EYE)を手動で指定する
.TP
\fB\-T\fR, \fB-\-tongue\fR\fI=TONGUE\fR
\&牛の舌に使う文字列(TONGUE)を手動で指定する
.TP
\fB\-f\fR, \fB-\-file\fR\fI=COWFILE\fR
\&使用するCOWFILEを指定する。
.br
\&規定のディレクトリおよび、環境変数COWPATHとして検索パスに指定されたディレクトリ内のCOWFILEであれば、
.br
\&拡張子(.cow)を含めない短い名前を用いることができる。
.br
\&相対パスまたは絶対パスで指定することで、COWPATH内に含まれないCOWFILEも指定可能
.TP
\fB\-R\fR, \fB-\-recursive\fR\fI=INT\fR
\&メッセージボックスをINTの回数だけ再帰的に描画する。
.TP
\fB\-l\fR, \fB-\-list\fR
\&COWPATHとして指定されたディレクトリ内のCOWFILEをリストアップして終了する
.TP
\fB-\-say\fR
.br
\&話すモード。cowsayと同じ動作。
.br
\&下記の\fI--think\fRが指定されない限り、標準で使用される
.TP
\fB-\-think\fR
.br
\&考えるモード。cowthinkと同じ動作
.TP
\fB-\-help\fR
.br
\&この使い方を表示して終了する
.TP
\fB\-\-version\fR
.br
\&バージョン情報を表示して終了する

.SS "終了ステータス:"
.TP
0
正常終了、
.TP
1
COW構造体の初期化に失敗、
.TP
2
COWFILEのオープンに失敗、
.TP
3
COWFILEの読み込みに失敗、
.TP
4
MSG構造体の初期化に失敗、
.TP
5
標準入力の読み出しに失敗、または引数として与えられた文字列の読み込みに失敗、
.TP
6
子プロセスの生成、またはパイプの生成に失敗、
.TP
7
正規表現のコンパイルに失敗、
.TP
8
\fI\-l, \-\-list\fRオプションでリストアップする際に、環境変数の分解に失敗、
.TP
255
引数の解析に失敗（例：無効な引数を渡した）

.SH 環境変数
COWPATH          COWFILEの検索パスを動的に指定する。
.br
                 この環境変数は、多くのシェル環境変数と同様にパスの複数記述が可能であり、優先順位に関する処理も同様である。

DEFAULT_COWFILE  標準で使用するCOWFILEを動的に指定する。
.br
                 通常、\fI\-\-file\fRオプションを使用せずに実行した場合、実行ファイルにビルドインされたCOWFILEを参照するが、
                 この環境変数が存在する場合、上記より優先して使用される。
                 \fI\-\-file\fRオプションと同様、COWPATH内に存在するCOWFILEであれば単体ファイル名及び、拡張子の省略記述も可能。

.SH 追加情報
.PP
ビルドする際、実行ファイルに対して静的に組み込まれるCOWPATHはMakefileの\fICOWPATH変数\fR、
.br
加えてCOWFILEは、src/config.hの\fIDEFAULT_COWFILE定数\fRにて変更できます。
.br
その他の既定動作に関しても、src/config.hの定数で変更できますので、
.br
色々と試してみるのも楽しいかもしれません。

.SH 謝辞
Masaya Tk (@844196)さんには、Homebrew用のリポジトリを提供して頂いた他、
.br
とても素敵なCOWFILEを沢山作って頂きました。
.br
日頃より参考になるアドバイスもして頂き、心より感謝を申し上げます。
.br
詳しくは、AUTHORSクレジットをご覧ください。

.SH 作者
sasairc (@sasairc)
.br
Masaya Tk (@844196) contributed to this project.

.SH 著作権
Copyright(c) 2015 sasairc <sasairc@ssiserver.moe.hm>
.br
This work is free. You can redistribute it and/or modify it under the
.br
terms of the Do What The Fuck You Want To Public License, Version 2,
.br
as published by Sam Hocevar.HocevarHocevar See the COPYING file or http://www.wtfpl.net/
.br
for more details.

.SH 関連項目
.B cowsay
.B renge
.B yasuna


src/clangsay.c
/*
 * clangsay -  The classic cowsay program, written in C.
 *
 * clangsay.c
 *
 * Copyright (c) 2015 sasairc
 * This work is free. You can redistribute it and/or modify it under the
 * terms of the Do What The Fuck You Want To Public License, Version 2,
 * as published by Sam Hocevar.HocevarHocevar See the COPYING file or http://www.wtfpl.net/
 * for more details.
 */

#include "./config.h"
#include "./clangsay.h"
#include "./cow.h"
#include "./msg.h"
#include "./info.h"
#include "./list.h"
#include <stdio.h>
#include <stdlib.h>
#include <getopt.h>

#ifdef  WITH_SHARED
#include <benly/string.h>
#else
#include <libbenly/src/string.h>
/* WITH_SHARED */
#endif

static void release(COW* cow, MSG* msg);

int main(int argc, char* argv[])
{
    int     res     = 0,
            index   = 0,
            rarg    = 0,
            status  = 0;

    char*   cowfile = NULL;

    COW*    cow     = NULL;

    MSG*    msg     = NULL;

    COWOPT  opt     = {
        COWOPT_ALLNO_FLAGS,
    };

    /* option for getopt_long() */
    struct  option opts[] = {
        {"eye",       required_argument, NULL, 'e'},
        {"tongue",    required_argument, NULL, 'T'},
        {"file",      required_argument, NULL, 'f'},
        {"recursive", required_argument, NULL, 'R'},
        {"nowrap",    no_argument,       NULL, 'n'},
        {"borg",      no_argument,       NULL, 'b'},
        {"dead",      no_argument,       NULL, 'd'},
        {"greedy",    no_argument,       NULL, 'g'},
        {"paranoid",  no_argument,       NULL, 'p'},
        {"stoned",    no_argument,       NULL, 's'},
        {"tired",     no_argument,       NULL, 't'},
        {"wired",     no_argument,       NULL, 'w'},
        {"youthful",  no_argument,       NULL, 'y'},
        {"list",      no_argument,       NULL, 'l'},
        {"say",       no_argument,       NULL,  0 },
        {"think",     no_argument,       NULL,  1 },
        {"help",      no_argument,       NULL,  2 },
        {"version",   no_argument,       NULL,  3 },
        {0, 0, 0, 0},
    };

    /* processing of arguments */
    while ((res = getopt_long(argc, argv, "nbdgpstwye:T:f:R:l", opts, &index)) != -1) {
        switch (res) {
            case    'e':
                opt.mode |= MODE_M_EYE;
                opt.eye= optarg;
                break;
            case    'T':
                opt.mode |= MODE_M_TONGUE;
                opt.tongue = optarg;
                break;
            case    'f':
                cowfile = optarg;
                break;
            case    'R':
                opt.mode |= MODE_MSG_RECURSIVE;
                if (strisdigit(optarg) < 0) {
                    fprintf(stderr, "%s: %s: invalid argument\n",
                            PROGNAME, optarg);
                    return -1;
                }
                rarg = atoi(optarg);
                break;
            case    'b':
                opt.mode |= MODE_BORG;
                break;
            case    'd':
                opt.mode |= MODE_DEAD;
                break;
            case    'g':
                opt.mode |= MODE_GREEDY;
                break;
            case    'p':
                opt.mode |= MODE_PARANOID;
                break;
            case    's':
                opt.mode |= MODE_STONED;
                break;
            case    't':
                opt.mode |= MODE_TIRED;
                break;
            case    'w':
                opt.mode |= MODE_WIRED;
                break;
            case    'y':
                opt.mode |= MODE_YOUTHFUL;
                break;
            case    'l':
                list_cowfiles();
            case    0:
                opt.mode |= MODE_SAY;
                break;
            case    1:
                opt.mode |= MODE_THINK;
                break;
            case    2:
                print_usage();
            case    3:
                print_version();
            case    '?':
                return -1;
        }
    }

    /* check env $DEFAULT_COWFILE */
    if (cowfile == NULL) {
        if ((cowfile = getenv("DEFAULT_COWFILE")) == NULL)
            cowfile = DEFAULT_COWFILE;
    }

    /* initialize cow */
    if (init_cow(&cow) < 0) {
        status = 1; goto ERR;
    }

    /* open cow file */
    if (cow->open(&cow, cowfile) < 0) {
        status = 2; goto ERR;
    }

    /* reading cow file to array */
    if (cow->read(&cow) < 0) {
        status = 3; goto ERR;
    }

    /* initialize msg */
    if (init_msg(&msg) < 0) {
        status = 4; goto ERR;
    }

    /* read argv or stdin */
    if (msg->read(&msg, argc, optind, argv) < 0) {
        status = 5; goto ERR;
    }

    /* <<<< recursive message box >>>> */
    if (opt.mode & MODE_MSG_RECURSIVE) {
        if (msg->recursive(&msg, rarg) < 0) {
            status = 6; goto ERR;
        }
    }

    /* print string */
    if (msg->print(msg) < 0) {
        status = 7; goto ERR;
    }

    /* print cow */
    cow->print(cow, &opt);

    /* release memory */
    release(cow, msg);

    return 0;

ERR:
    release(cow, msg);

    return status;
}

static
void release(COW* cow, MSG* msg)
{
    if (cow != NULL)
        cow->release(cow);

    if (msg != NULL)
        msg->release(msg);

    return;
}
